pkg/filesystem/path/windows_parser.go

```diff
@@ -0,0 +1,152 @@
+package path
+
+import (
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type windowsParser struct {
+	path string
+}
+
+type windowsRelativeParser struct {
+	path string
+}
+
+// NewWindowsParser creates a Parser for Windows paths that can be used in Resolve.
+func NewWindowsParser(path string) (Parser, error) {
+	// Windows paths are generally passed to system calls that
+	// accept C strings. There is no way these can accept null
+	// bytes.
+	if strings.ContainsRune(path, '\x00') {
+		return nil, status.Error(codes.InvalidArgument, "Path contains a null byte")
+	}
+	return &windowsParser{path}, nil
+}
+
+// MustNewWindowsParser is identical to NewWindowsParser, except that it panics
+// upon failure.
+func MustNewWindowsParser(path string) Parser {
+	parser, err := NewWindowsParser(path)
+	if err != nil {
+		panic(err)
+	}
+	return parser
+}
+
+func (p windowsParser) ParseScope(scopeWalker ScopeWalker) (next ComponentWalker, remainder RelativeParser, err error) {
+	path := strings.Replace(p.path, "/", "\\", -1)
+
+	if len(path) >= 3 {
```

Comment:
This ends up interpreting paths like `C:` and `C:Hello.txt` as relative paths.

Even though `C:` and `C:Hello.txt` technically speaking have a different meaning than `C:\` and `C:\Hello.txt`, I think we should simply interpret them as such.

I suspect the right thing to do is to only consume the first two characters, and then call something similar to `stripOneOrMoreSlashes()`.

Reply:
What about this?

```go
if len(path) >= 2 && (
    (path[0] >= 'A' && path[0] <= 'Z') ||
    (path[0] >= 'a' && path[0] <= 'z')
) && path[1] == ':' {
    next, err = scopeWalker.OnDriveLetter(rune(path[0]&0x1f + 'A' - 1))
    if err != nil {
        return nil, nil, err
    }
    return next, windowsRelativeParser{stripOneOrMoreSlashes(path[2:])}, nil
}
```

Reply:
Looks good. The pathological cases do contain invalid characters, so we will error out if they are used rather than attempting to use files that it should not use. So I think this is fine.